### DISCUSSION TITLE
How to add additional fetchOption parameters without overwriting client default

### DISCUSSION THREAD

**Discussion by Svarto** [urql-graphql/urql#1477] [Q&A] [answered]
Hi,

I have a mutation where I want to send a session variable to do some backend checks, I can do so rather easily by setting the fetchOptions in the context - however this overwrites my client defaults that I have set up. Is there any way to write the mutation so that it just adds to the headers already created by the defaults (i.e. since it is a one off, I want it on top of the default).

```
      executeMutation(
        { email: email },
        {
          fetchOptions: () => {
            return { headers: { "x-hasura-email-check": mail } };
          },
        }
      )
```

When I execute the above mutation, it overwrites my client defaults which is not what I want. I still want the client defaults and just add the above "on top". Hope it makes sense? Thank you for all and any help!!

Client defaults:
```
export const client = createClient({
  exchanges: [devtoolsExchange, dedupExchange, CacheExchange, fetchExchange],
  url: graphqlUrl,
  fetchOptions: () => {
    const token = auth.getJWTToken();
    return {
      headers: { authorization: token ? `Bearer ${token}` : "" },
    };
  },
});
```

**Comment by kitten** [comment, marked as answer]
I can't really find a reference to `x-hasura-email-check`, so if that's a custom addition, I'd consider putting it onto an actual variable, but then again, I can only assume what it may be 😅 

There isn't really a way to compose it in this manner unless you'd use a custom exchange to merge different options, e.g. `authToken` and `emailCheck` into `fetchOptions`. Alternatively you could use the `authExchange` to put the `authorization` into a different place, which would be the quickest way to fix this.

But a custom exchange could look like the following:

```js
import { makeOperation } from '@urql/core';

export const fetchOptionsExchange: Exchange = ({ forward }) => ops$ => {
  return pipe(
    ops$,
    map((operation: Operation) => {
      return makeOperation(operation.kind, operation, {
        ...operation.context,
        fetchOptions: {
          ...(typeof operation.context.fetchOptions === 'function'
            ? operation.context.fetchOptions()
            : operation.context.fetchOptions),
          ...operation.context.extraFetchOptions // whatever you need here
        },
      });
    }),
    forward
  );
};
```
